File: website/vo2-go/geo.go
package vo2

import "math"

func toRadians(deg float64) float64 { return deg * math.Pi / 180.0 }

func haversine(lat1, lon1, lat2, lon2 float64) float64 {
	const r = 6_371_000.0
	dlat := toRadians(lat2 - lat1)
	dlon := toRadians(lon2 - lon1)
	sLat := math.Sin(dlat / 2)
	sLon := math.Sin(dlon / 2)
	a := sLat*sLat + math.Cos(toRadians(lat1))*math.Cos(toRadians(lat2))*sLon*sLon
	c := 2 * math.Atan2(math.Sqrt(a), math.Sqrt(1-a))
	return r * c
}

File: website/vo2/kt/src/wasmJsMain/kotlin/Geo.kt
import kotlin.math.PI
import kotlin.math.atan2
import kotlin.math.cos
import kotlin.math.sin
import kotlin.math.sqrt

private const val EARTH_RADIUS_M = 6_371_000.0

private fun Double.toRadians(): Double = this * PI / 180.0

internal fun haversine(lat1: Double, lon1: Double, lat2: Double, lon2: Double): Double {
    val dLat = (lat2 - lat1).toRadians()
    val dLon = (lon2 - lon1).toRadians()
    val sLat = sin(dLat / 2)
    val sLon = sin(dLon / 2)
    val a = sLat * sLat + cos(lat1.toRadians()) * cos(lat2.toRadians()) * sLon * sLon
    val c = 2 * atan2(sqrt(a), sqrt(1 - a))
    return EARTH_RADIUS_M * c
}
